feat(intl): ignore script subtags when canonicalizing locale strings

kszczek opened the pull request:
Locale strings with a script such as `zh-Hans-CN` are incorrectly canonicalized as `zh` instead of `zh-CN`. Update the `canonicalizedLocale` function to look for a script subtag and ignore it if found.

Closes: https://github.com/dart-lang/i18n/issues/945

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/blob/main/docs/External-Package-Maintenance.md#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

mosuem commented on pkgs/intl/lib/src/intl_helpers.dart:
```diff
@@ -165,6 +178,9 @@ String canonicalizedLocale(String? aLocale) {
   }
   var language = aLocale.substring(0, separatorIndex);
   var region = aLocale.substring(separatorIndex + 1);
```
This is not really the `region` anymore, more something like `subtags`, right? Which we then try to split into `script` and `region`.

kszczek replied:
Correct, I've refactored this to make the distinction.

mosuem reviewed (changes requested):
A bit of a hacky solution, but the existing code is also quite hacky, so that's fine :)

mosuem commented:
This fails internal Google testing, which looks like this:

```
test('Recover locale from string', () {
      expect(
        Locale.tryParse(Intl.canonicalizedLocale('zh_Hant_TW')),
        Locale.fromSubtags(
          languageCode: 'zh',
          scriptCode: 'Hant',
          countryCode: 'TW',
        ),
      );
    });
```

Which made me think about what `canonicalizedLocale` promises vs does. The documentation states it turns a locale string into the form `xx_YY`, but also `... where it might possibly be in the wrong case or with a hyphen instead of an underscore.`.
So it seems it was interpreted as trying to do [canonicalization](https://unicode-org.github.io/icu/userguide/locale/#canonicalization), more precisely level 1 canonicalization.

It doesn't help that `NumberFormat.currency` can't handle scripts... I think we should remedy that instead. This would mean that at https://github.com/dart-lang/i18n/blob/c6b911b2312ba449df783684a451f406afafc75c/pkgs/intl/lib/src/intl_helpers.dart#L188, we insert another fallback after `canonicalizedLocale` but before `shortLocale` (which should be renamed to something like `languageOnly`), which returns what you implemented. This way, we would first try `zh_Hant_CN`, then `zh_CN`, then `zh`, etc...

What do you think?

mosuem reviewed (changes requested):
As detailed in comment

mosuem commented:
I would also add something like 

```dart
  test('Use script', () {
    expect(
      NumberFormat.currency(locale: 'zh_Hant_TW').format(12),
      'TWD12.00',
    );
    expect(
      NumberFormat.currency(locale: 'zh_Hant_CN').format(12),
      'CNY12.00',
    );
  });
```

to the tests.

kszczek commented:
Thank you for your patience, I couldn't get back to this sooner.

> we insert another fallback after canonicalizedLocale but before shortLocale (which should be renamed to something like languageOnly), which returns what you implemented. This way, we would first try zh_Hant_CN, then zh_CN, then zh, etc. when running NumberFormat.currency.

Sounds good to me, even though there is no script specific data for numbers nor dates as far as I know, we should still try looking them up in case they appear in the future.

Implemented your suggestions and modified tests, let me know if I should change anything else.
